DSDEGP-1639: better upsert id type

pshapiro4broad opened the pull request:
### Description

DSDEGP-1639: create a type for upsert IDs
- upsert ID is now of type UpsertId
- refactored code that creates a filename from the upsert ID

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad committed:
DSDEGP-1639: create a type for upsert IDs

- upsert ID is now of type UpsertId
- refactored code that creates a filename from the upsert ID

pshapiro4broad committed:
Merge remote-tracking branch 'origin/develop' into ps_DSDEGP-1639_better_upsert_id_type

danxmoran commented on clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ClioDocument.scala:
```diff
@@ -23,6 +25,13 @@ trait ClioDocument {
+    *
+    * @return the filename where this document's upsert data is stored
+    */
+  def persistenceFilename: String = ClioDocument.persistenceFilename(upsertId)
```
I don't think we need this, since it just forwards to the method on the companion object (which is visible everywhere).

pshapiro4broad replied:
If I remove it, how would `writePath.resolve(s"${document.persistenceFilename}")` work?

danxmoran replied:
`writePath.resolve(s"${ClioDocument.persistenceFilename(document.upsertId)}")`

danxmoran commented on clio-util/src/main/scala/org/broadinstitute/clio/util/model/UpsertId.scala:
```diff
@@ -0,0 +1,6 @@
+package org.broadinstitute.clio.util.model
+
+sealed case class UpsertId(id: String) extends Ordered[UpsertId] {
```
I think we want `final` instead of `sealed`.

pshapiro4broad replied:
ok

pshapiro4broad committed:
DSDEGP-1639: code review feedback

- make callers use the ClioDocument.persistenceFilename getter
- make case class final, not sealed

danxmoran commented on clio-util/src/main/scala/org/broadinstitute/clio/util/json/ModelAutoDerivation.scala:
```diff
@@ -61,6 +63,16 @@ trait ModelAutoDerivation extends AutoDerivation {
+    for {
+      upsertId <- c.as[String]
+    } yield {
+      UpsertId(upsertId)
```
Are there any properties about the `upsertId` that this should check? Length would be an easy one.

pshapiro4broad replied:
Yeah, it would make sense to do this, inside `UpsertId` or `UpsertIdGenerator`.

tbl3rd commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/PersistenceDAO.scala:
```diff
@@ -111,31 +111,31 @@ trait PersistenceDAO extends LazyLogging {
     * @return a Future Seq of ClioDocument
     */
-  def getAllSince[D <: ClioDocument: Decoder](upsertId: Option[String],
+  def getAllSince[D <: ClioDocument: Decoder](document: Option[ClioDocument],
```
Good idea.

tbl3rd commented on clio-server/src/main/scala/org/broadinstitute/clio/server/service/PersistenceService.scala:
```diff
@@ -30,14 +31,14 @@ class PersistenceService private (persistenceDAO: PersistenceDAO,
     * @tparam TM The type of the Transfer Metadata DTO.
     * @tparam D  The type of the Document.
-    * @return A future result of the upsert.
+    * @return A future result of the ID for this upsert.
```
How about "a future ID for this upsert" or just "an ID for this upsert" ?

pshapiro4broad replied:
ok

pshapiro4broad committed:
DSDEGP-1639: code review feedback

- merge UpsertId and UpsertIdGenerator
- restrict access to UpsertId creation
- update comment
